Foundation/NSLock.swift

```diff
@@ -258,3 +302,59 @@ open class NSCondition: NSObject, NSLocking {
     
     open var name: String?
 }
+
+private func timeSpecFrom(date: Date) -> timespec? {
+    let interval = date.timeIntervalSince1970
+    guard interval >= 0 else {
+        return nil
+    }
+
+    var ts = timespec()
+    ts.tv_sec = Int(interval.rounded(.down))
+    ts.tv_nsec = Int((interval - Double(ts.tv_sec)) * 1_000_000_000.0)
+    return ts
+}
+
+#if os(OSX) || os(iOS)
+
+private func allocateTimedLockData() -> UnsafeMutablePointer<_timed_lock_data>? {
+    let tld = UnsafeMutablePointer<_timed_lock_data>.allocate(capacity: 1)
+    if pthread_mutex_init(_CFGetTimedLockMutex(tld), nil) == 0 {
+        if pthread_cond_init(_CFGetTimedLockCond(tld), nil) == 0 {
+            return tld
+        } else {
+            pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
+        }
+    }
+    tld.deallocate(capacity: 1)
+    return nil
+}
+
+private func deallocateTimedLockData(_ tld: UnsafeMutablePointer<_timed_lock_data>?) {
+    if let tld = tld {
+        pthread_mutex_destroy(_CFGetTimedLockMutex(tld))
+        pthread_cond_destroy(_CFGetTimedLockCond(tld))
+        tld.deallocate(capacity: 1)
+    }
+}
+
+// Emulate pthread_mutex_timedlock using pthread_cond_timedwait.
+// lock(before:) has obtained a shared pointer to a _time_lock_data structure
+// and the condition variable to be waited on.
+// unlock() will check for the existance of this structure and if its pointer is
+// non-nil, it will use pthread_cond_broadcast() to stop any waits in progress.
+private func timedLock(mutex: _PthreadMutexPointer,
+                       endTime: inout timespec,
+                       using tld: UnsafeMutablePointer<_timed_lock_data>) -> Bool {
+    pthread_mutex_lock(_CFGetTimedLockMutex(tld))
+    let status = pthread_cond_timedwait(_CFGetTimedLockCond(tld),
+                                        _CFGetTimedLockMutex(tld),
+                                        &endTime)
+    pthread_mutex_unlock(_CFGetTimedLockMutex(tld))
```

Comment:
wait, why are we unlocking here?

Reply:
`pthread_cond_timedwait` requires the lock so it needs to be unlocked.

Reply:
@spevans sorry I think I'm slow. So I agree that `pthread_cond_timedwait` needs the lock to be locked which is the case. But the function is called `timedLock` so when the function returns, the lock should still be locked, right? Otherwise the function returns and the lock isn't actually locked.